ngraph/core/src/op/einsum.cpp

```diff
@@ -0,0 +1,278 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <algorithm>
+#include <cctype>
+#include <ngraph/validation_util.hpp>
+#include <regex>
+#include <string>
+#include <unordered_map>
+
+#include "itt.hpp"
+#include "ngraph/op/einsum.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+NGRAPH_RTTI_DEFINITION(op::v7::Einsum, "Einsum", 7);
+
+op::v7::Einsum::Einsum(const OutputVector& inputs, const std::string& equation)
+    : Op(inputs)
+    , m_equation(equation)
+{
+    constructor_validate_and_infer_types();
+}
+
+bool op::v7::Einsum::parse_equation(std::vector<std::string>& input_subscripts,
+                                    std::string& output_subscript)
+{
+    NGRAPH_OP_SCOPE(v7_Einsum_parse_equation);
+
+    // create regular expression to match subscript that consists of either alphabetical letters
+    // only or alphabetical letters with one ellisis
+    const std::regex subscript_regex("[A-Za-z]*|[A-Za-z]*\\.\\.\\.[A-Za-z]*");
+
+    // create regular expression to match ellipsis label
+    const std::regex ellipsis_regex("\\.\\.\\.");
+
+    // remove extra white-spaces from the equation
+    m_equation.erase(std::remove_if(m_equation.begin(), m_equation.end(), ::isspace),
+                     m_equation.end());
+
+    // split equation to input subscripts and an output subscript
+    auto pos_output_delimeter = m_equation.find("->");
+    auto input_subscripts_str = m_equation.substr(0, pos_output_delimeter);
+
+    // split the input subscripts into a vector of input subscripts
+    bool is_ellipsis_met = false;
+    std::smatch ellipsis_match;
+    input_subscripts.clear();
+    std::istringstream input;
+    input.str(input_subscripts_str);
+    for (std::string input_subscript; std::getline(input, input_subscript, ',');)
+    {
+        // check that input subscript contains only alphabetic letter or ellipsis
+        if (std::regex_match(input_subscript, subscript_regex) == false)
+        {
+            return false;
+        }
+
+        // figure out if ellipsis is met in input subscript
+        if (std::regex_search(input_subscript, ellipsis_match, ellipsis_regex) == true)
+        {
+            is_ellipsis_met = true;
+        }
+        input_subscripts.push_back(input_subscript);
+    }
+
+    if (pos_output_delimeter == std::string::npos)
+    {
+        // recover output subscript
+        output_subscript = "";
+        for (auto const& input_subscript : input_subscripts)
+        {
+            for (auto const& label : input_subscript)
+            {
+                if (std::isalpha(label) && output_subscript.find(label) == std::string::npos)
+                {
+                    output_subscript += label;
+                }
```

Comment:
nit: It would be just a little bit faster to put all labels into a set (no need to check `output_subscript.find(label) == std::string::npos`) and after the `for` loop just to create `output_subscript` from the set.

Reply:
It seems I have to use one more loop for transferring data between `set` and `string`. Due to relative small size of equation I think only small benefit may exist or not.